src/ephemeris.c

```diff
@@ -68,6 +70,129 @@ static s8 calc_sat_state_xyz(const ephemeris_t *e, const gps_time_t *t,
+            + acc[0];
+
+  ydot[4] = -GPS_GM / pow(r,3) * pos[1]
+            - 3/2 * GLO_J02 * GPS_GM * pow(GLO_A_E, 2) / pow(r, 5) *
+              pos[1] * (1 - 5 * pow(pos[2], 2) / pow(r, 2))
+            + pow(GPS_OMEGAE_DOT, 2) * pos[1]
+            + 2 * GPS_OMEGAE_DOT * vel[0]
+            + acc[1];
+
+  ydot[5] = -GPS_GM / pow(r,3) * pos[2]
+            - 3/2 * GLO_J02 * GPS_GM * pow(GLO_A_E, 2) / pow(r, 5) *
+              pos[2] * (3 - 5 * pow(pos[2], 2) / pow(r, 2))
+            + acc[2];
+}
+
+/** Calculate satellite position, velocity and clock offset from GLO ephemeris.
+ *
+ * \param e Pointer to an ephemeris structure for the satellite of interest
+ * \param t time at which to calculate the satellite state
+ * \param pos Array into which to write calculated satellite position [m]
+ * \param vel Array into which to write calculated satellite velocity [m/s]
+ * \param clock_err Pointer to where to store the calculated satellite clock
+ *                  error [s]
+ * \param clock_rate_err Pointer to where to store the calculated satellite
+ *                       clock error [s/s]
+ *
+ * \return  0 on success,
+ *         -1 if ephemeris is not valid or too old
+ */
+static s8 calc_sat_state_glo(const ephemeris_t *e, const gps_time_t *t,
+                             double pos[3], double vel[3],
+                             double *clock_err, double *clock_rate_err)
+{
+  assert(e != NULL);
+  assert(t != NULL);
+  assert(pos != NULL);
+  assert(vel != NULL);
+  assert(clock_err != NULL);
+  assert(clock_rate_err != NULL);
+
+  double dt = abs(gpsdifftime(t, &e->toe));
+
+  if (dt > 900) {
+    log_error("GLO: Integration end point is not within 900 s of TOE");
+    return 1;
+  }
+
+  if (ceil(dt / GLO_NUM_STEPS)) {
+    double ydot0[6], yn[6];
+    double h = dt / GLO_NUM_STEPS;
+
+    calc_ydot(ydot0, e->glo.pos, e->glo.vel, e->glo.acc);
+    memcpy(&yn[0], e->glo.pos, sizeof(double) * 3);
+    memcpy(&yn[3], e->glo.vel, sizeof(double) * 3);
+
+    /* Runge-Kutta integration algorithm */
+    for (u8 i = 0; i < GLO_NUM_STEPS; i++) {
+      double k1[6], k2[6], k3[6], k4[6], y_tmp[6];
+      u8 j;
+
+      memcpy(k1, ydot0, sizeof(k1));
+
+      for (j = 0; j < 6; j++)
+        y_tmp[j] = yn[j] + h/2 * k1[j];
+
+      calc_ydot(k2, &y_tmp[0], &y_tmp[3], e->glo.acc);
+
+      for (j = 0; j < 6; j++)
+        y_tmp[j] = yn[j] + h/2 * k2[j];
+
+      calc_ydot(k3, &y_tmp[0], &y_tmp[3], e->glo.acc);
+
+      for (j = 0; j < 6; j++)
+        y_tmp[j] = yn[j] + h * k3[j];
+
+      calc_ydot(k4, &y_tmp[0], &y_tmp[3], e->glo.acc);
+
+      for (j = 0; j < 6; j++)
+        yn[j] += h/6 * (k1[j] + k2[j] + k3[j] + k4[j]);
+
+      calc_ydot(ydot0, &yn[0], &yn[3], e->glo.acc);
+    }
+    memcpy(pos, &yn[0], sizeof(double) * 3);
+    memcpy(vel, &yn[3], sizeof(double) * 3);
+  } else {
+    memcpy(pos, e->glo.pos, sizeof(double) * 3);
+    memcpy(vel, e->glo.vel, sizeof(double) * 3);
+  }
+
+  *clock_err = e->glo.tau;
```

Comment:
I checked another open source implementation of the clock correction in GPSTk, and they don't do iterations.

Basically the equation should be:
`*clock_err = e->glo.tau + e->glo.gamma * dt;`